### PR TITLE
Allow Type extension

### PR DESCRIPTION
**Pull request by rhelms**
Does not implement allOf in JSON Schema
encode/apistar#435

### CODE REVIEW

**Comment by tomchristie** [comment]
Great stuff. I'd prefer that we don't introduce so much extra code into the test cases, as I think it's unnecessarily verbose there. I'll comment on those cases inline...

File: tests/test_types.py [hunk 1]
```diff
@@ -35,6 +39,23 @@ def test_from_object():
     }
 
 
+def test_extended_from_object():
```
Comment (tomchristie) [inline tests/test_types.py]: Yup this is reasonable.

File: tests/test_types.py [hunk 2]
```diff
@@ -49,6 +70,22 @@ def test_from_kwargs():
     }
 
 
+def test_extended_from_kwargs():
```
Comment (tomchristie) [inline tests/test_types.py]: Let's have this one too.

File: tests/test_types.py [hunk 3]
```diff
@@ -63,6 +100,22 @@ def test_from_dict():
     }
 
 
+def test_extended_from_dict():
```
Comment (tomchristie) [inline tests/test_types.py]: Okay, for consistency.

File: tests/test_types.py [hunk 4]
```diff
@@ -98,6 +151,59 @@ def test_setattr():
     assert exc.value.detail == 'May not be null.'
 
 
+def test_extended_setattr():
```
Comment (tomchristie) [inline tests/test_types.py]: I think this becomes unnecessary.

File: tests/test_types.py [hunk 5]
```diff
@@ -133,6 +239,59 @@ def test_setitem():
     assert exc.value.detail == 'May not be null.'
 
 
+def test_extended_setitem():
```
Comment (tomchristie) [inline tests/test_types.py]: Again, let's drop this one.

File: tests/test_types.py [hunk 6]
```diff
@@ -150,6 +309,23 @@ def test_misc():
         Product([])
 
 
+def test_extended_misc():
```
Comment (tomchristie) [inline tests/test_types.py]: Let's drop this too. Test cases are a maintenance cost too.

File: tests/test_types.py [hunk 7]
```diff
@@ -187,3 +363,90 @@ def test_as_jsonschema():
...
+                    }
+                },
+                "required": [
+                    "reviewer"
+                ]
+            }
+        }
+    }
+
+    """
```
Comment (tomchristie) [inline tests/test_types.py]: I'd suggest we drop the docstring here. The behavior's currently good as it is, we don't need to document what we may intend to do in the future.

**Comment by rhelms** [comment]
Review changes have been committed.

**Comment by tomchristie** [comment]
Excellent, thank you!

**Pull request merged**